File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/viewmodel/BaseViewModel.java
package openfoodfacts.github.scrachx.openfood.views.viewmodel;

import androidx.annotation.NonNull;

import io.reactivex.disposables.CompositeDisposable;

/**
 * Created by Abdelali Eramli on 27/12/2017.
 */
public abstract class BaseViewModel {
    protected CompositeDisposable subscriptions;

    public BaseViewModel() {
    }

    public void bind() {
        unbind();
        subscriptions = new CompositeDisposable();
        subscribe(subscriptions);
    }

    public void unbind() {
        if (subscriptions != null) {
            subscriptions.clear();
            subscriptions = null;
        }
    }

    protected abstract void subscribe(@NonNull final CompositeDisposable subscriptions);


}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/viewmodel/BaseViewModel.kt
package openfoodfacts.github.scrachx.openfood.features.viewmodel

import androidx.lifecycle.ViewModel
import io.reactivex.disposables.CompositeDisposable

/**
 * Created by Abdelali Eramli on 27/12/2017.
 */
abstract class BaseViewModel : ViewModel() {
    var subscriptions: CompositeDisposable? = null
    fun bind() {
        unbind()
        CompositeDisposable().also {
            subscriptions = it
            subscribe(it)
        }

    }

    fun unbind() {
        subscriptions?.let {
            it.clear()
            subscriptions = null
        }
    }

    protected abstract fun subscribe(subscriptions: CompositeDisposable)
}